File: battery-modeling/Step 3. Simulation/mmaptx.m
classdef Mmaptx < handle
    % Matlab structure for memmap transfering with python
    %   Class that enables memory mapped files data transfer method with
    %   the respective python library.
    %
    % WARNING: This was slower than direct implementation. It's better to
    % use the memory map directly instead through this interface if speed
    % is important.

    properties
        mm_in
        mm_out
        blocking
        sync = 0;
    end

    methods
        function obj = Mmaptx(name, ftype, blocking)
            %Mmaptx Construct an instance of this class
            %   Initializes the memory mapped files for transfer and the
            %   class parameters/properties

            obj.mm_in = memmapfile(strcat(name,'_mmap_out.dat'),'Format', ftype);

            obj.mm_out = memmapfile(strcat(name,'_mmap_in.dat'),'Format', ftype);
            obj.mm_out.Writable = true;
            obj.blocking = blocking;
        end

        function write(obj, input)
            %METHOD1 Summary of this method goes here
            %   Detailed explanation goes here
            obj.mm_out.Data(2:end) = input;
            obj.mm_out.Data(1) = obj.sync;
            
        end

        function data = read(obj)
            %read Reads data from the memory mapped file
            %   Read data from the memory mapped file
            %   If blocking is enabled, it will read only when the sync
            %   flag changes in the input file.
            %   Returns the read data and updates the sync flag for sending

            if obj.blocking == true
                while obj.mm_in.Data(1) == obj.sync; end % blocking wait
            end
            obj.sync = obj.mm_in.Data(1);
            data = obj.mm_in.Data(2:end);
        end
    end
end

File: battery-modeling/Step 3. Simulation/mmapTx.py
import mmap
import struct
import atexit
import os

# Library that can communicate with other programs through memory mapped files
# It writes to an output memory mapped file and it reads from a different input memory mapped file
# It can also sync data with the other program via a sync variable that is passed to the other
# program and input data will only be read unless that flag is sent back in the input memory map file (if blocking read is enabled)
class Mmaptx:

    def __init__(self, name = None, format_type = "d", in_size = 1, out_size = 1, blocking = True):

        atexit.register(self.cleanup)

        # Save configuration data
        self.ftype = format_type
        self.ftype_size = struct.calcsize(format_type)
        self.blocking = blocking
        self.in_size = in_size
        self.out_size = out_size

        if name == None:
            exit("Name must be specified.")
        
        self.filename = name
        self.filename_in = name + "_mmap_in.dat"
        self.filename_out = name + "_mmap_out.dat"

        # Create or clean files that will be used as memory map input and output
        f = open(self.filename_in, "w+b")
        f.write(struct.pack(self.ftype * (self.in_size+1), *([0]*(self.in_size+1))))
        f.flush()
        f.close()

        f = open(self.filename_out, "w+b")
        f.write(struct.pack(self.ftype * (self.out_size+1), *([0]*(self.out_size+1))))
        f.flush()
        f.close()

        del f

        # Crete the memory maps
        self.f_in = open(self.filename_in, "r+b")
        self.mm_in = mmap.mmap(self.f_in.fileno(), 0)

        self.f_out = open(self.filename_out, "a+b")
        self.mm_out = mmap.mmap(self.f_out.fileno(), 0)

        self.sync = 0.0
    
    # Write to output memmap
    def write(self, *data):
        
        out_size_l = len(data)
        self.mm_out.seek(self.ftype_size)
        self.mm_out.write(struct.pack(self.ftype * out_size_l, *data))
        
        self.sync = self.sync
        self.mm_out.seek(0)
        self.mm_out.write(struct.pack(self.ftype, self.sync))

    # Blocking read from input memmap
        # Waits until the other program has set the sync flag to the same value sent
        # This means data is ready for reading
    def read(self):

        self.mm_in.seek(0)
        sync_in = struct.unpack(self.ftype, self.mm_in.read(self.ftype_size))

        if self.blocking:
            while self.sync == sync_in[0]:
                self.mm_in.seek(0)
                sync_in = struct.unpack(self.ftype, self.mm_in.read(self.ftype_size))
            self.sync = sync_in[0]

        data_in = struct.unpack(self.ftype * self.in_size, self.mm_in.read(self.ftype_size * self.in_size))
            
        return data_in
    
    # Close files and memmaps
    def cleanup(self):

        self.mm_out.close()
        self.f_out.close()
        self.mm_in.close()
        self.f_in.close()
